File: backend/app/routers/wheel.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.database import get_db
from app.models.user import User
from app.schemas.stars import StarsInvoiceCreateOut, StarsInvoiceStatusOut
from app.schemas.wheel import WheelSpinResultOut, WheelStatusOut
from app.services import stars_payment_service, wheel_service

router = APIRouter(prefix="/wheel", tags=["wheel"])


@router.get("/status", response_model=WheelStatusOut)
async def get_status(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await wheel_service.get_status(db, user)


@router.post("/spin/free", response_model=WheelSpinResultOut)
async def spin_free(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await wheel_service.spin_free(db, user)


@router.post("/spin/coins", response_model=WheelSpinResultOut)
async def spin_paid_coins(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await wheel_service.spin_paid_coins(db, user)


@router.post("/spin/stars-invoice", response_model=StarsInvoiceCreateOut)
async def create_stars_spin_invoice(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await wheel_service.create_spin_invoice(db, user)


@router.get("/stars-invoices/{payload_token}", response_model=StarsInvoiceStatusOut)
async def get_stars_spin_invoice_status(
    payload_token: str, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    return await stars_payment_service.get_invoice_status(db, user, payload_token)

File: frontend/src/api/wheel.ts
import { api } from "@/lib/api";
import type { StarsInvoiceCreate, StarsInvoiceStatus, WheelSpinResult, WheelStatus } from "@/types";

export async function fetchWheelStatus(): Promise<WheelStatus> {
  const { data } = await api.get<WheelStatus>("/wheel/status");
  return data;
}

export async function spinFree(): Promise<WheelSpinResult> {
  const { data } = await api.post<WheelSpinResult>("/wheel/spin/free");
  return data;
}

export async function spinPaidCoins(): Promise<WheelSpinResult> {
  const { data } = await api.post<WheelSpinResult>("/wheel/spin/coins");
  return data;
}

export async function createWheelStarsInvoice(): Promise<StarsInvoiceCreate> {
  const { data } = await api.post<StarsInvoiceCreate>("/wheel/spin/stars-invoice");
  return data;
}

export async function fetchWheelStarsInvoiceStatus(payloadToken: string): Promise<StarsInvoiceStatus> {
  const { data } = await api.get<StarsInvoiceStatus>(`/wheel/stars-invoices/${payloadToken}`);
  return data;
}
